File: Linux/java/prime3a.java
import java.io.*;

public class prime3a {
    public static void main(String[] args) {
        int number = 1;
        int count;
        int divisor;
        long divisorSquare;
        int remainder;
        boolean isprime;
        int maxPrimes = 100;
        boolean printToFile=false;

        try {
            if (args.length > 0) {
                maxPrimes=Integer.parseInt(args[0]);
            }
        } catch (Exception e) {
            System.err.println(String.format("Error parsing count:%n%s%n", e.toString()));
        }
        try {
            if (args.length > 1) {
                PrintStream out = new PrintStream(new FileOutputStream(args[1]));
                System.setOut(out);
                printToFile=true;
            }
        } catch (Exception e) {
            System.err.println(String.format("Error parsing output filename:%n%s%n", e.toString()));
        }

        System.out.println(printPrime(1, 2));
        number = 1;
        for (count = 1; count < maxPrimes; count++) {
            do {
                number = 2 + number;
                divisor = 3;
                divisorSquare = divisor * divisor;
                isprime = true;
                while (isprime && divisorSquare <= number) {
                    remainder = number % divisor;
                    if (remainder == 0) {
                        isprime = false;
                    } else {
                        divisor = 2 + divisor;
                        divisorSquare = divisor * divisor;
                    }
                }
                if (isprime) {
                    System.out.println(printPrime(1 + count, number));
                }
            } while (!isprime);

            if (count % 1000 == 0) {
                System.err.println(String.format("%d primes found.", count));
            }
        }
        System.err.println(String.format("%d primes found.", maxPrimes));

        if (printToFile) {
            System.out.close();
        }
    }
    private static String printPrime(int seq, int number) {
        return String.format("%20s: %20s", seq, number);
    }
}

File: Linux/C/prime3a.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define OUTPUT_FILE_LEN 256
#define OUTPUT_FORMAT "%20i: %20i\n"

int main(int argc, char *argv[]) {
    unsigned int number = 1;
    unsigned int count;
    unsigned int num_primes = 100;
    unsigned int divisor;
    unsigned long divisor_square;
    unsigned int remainder;
    int isprime = 0;
    char output_file[OUTPUT_FILE_LEN];
    FILE *fp=stdout;

    if (argc > 1) {
        num_primes = (unsigned int)atoi(argv[1]);

        if (num_primes < 1) {
            num_primes = 100;
        }

        if (argc > 2) {
            strncpy(output_file, argv[2], OUTPUT_FILE_LEN - 1);

            if ((fp = fopen(output_file, "w")) == NULL) {
                fprintf(stderr, "Error opening %s\n", output_file);
                return 1;
            }
        }
    }

    fprintf(fp, OUTPUT_FORMAT, 1, 2);

    for (count = 1; count < num_primes; count++) {
        while (1) {
            number += 2;
            divisor = 3;
            divisor_square = divisor * divisor;
            isprime = 1;

            while (divisor_square <= number) {
                remainder = number % divisor;

                if (!remainder) {
                    isprime = 0;
                    break;
                } else {
                    divisor += 2;
                    divisor_square = divisor * divisor;
                }
            }
            if (isprime) {
                fprintf(fp, OUTPUT_FORMAT, 1+count, number);
                break;
            }
        }
        if (!(count % 1000)) {
            fprintf(stderr, "%i primes found\n", count);
        }
    }

    fclose(fp);
    fprintf(stderr, "%i primes found\n", count);
    return 0;
}